File: app/src/main/java/net/osdn/gokigen/pkremote/calendar/TargetMonthSetDialog.java
package net.osdn.gokigen.pkremote.calendar;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.NumberPicker;
import android.widget.TextView;

import net.osdn.gokigen.pkremote.R;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;

/**
 *    年と月を開くダイアログ
 *
 */
public class TargetMonthSetDialog extends DialogFragment
{
    private final String TAG = toString();
    private static final int YEAR_LIMIT_MIN = 2010;
    private static final int YEAR_LIMIT_MAX = 2050;
    private static final int MONTH_LIMIT_MIN = 1;
    private static final int MONTH_LIMIT_MAX = 12;

    private String title = "";
    private int yearNum = 0;
    private int monthNum = 0;
    private Callback callback = null;
    private Dialog myDialog = null;

    /**
     *
     *
     */
    public static TargetMonthSetDialog newInstance(String title, int year, int month, Callback callback)
    {
        TargetMonthSetDialog instance = new TargetMonthSetDialog();
        instance.prepare(title, year, month, callback);

        // パラメータはBundleにまとめておく
        Bundle arguments = new Bundle();
        arguments.putString("title", title);
        //arguments.putString("message", message);
        instance.setArguments(arguments);

        return (instance);
    }

    /**
     *
     *
     */
    private void prepare(String title, int yearNum, int monthNum, Callback callback)
    {
        this.title = title;
        this.yearNum = yearNum;
        this.monthNum = monthNum;
        this.callback = callback;
    }

    /**
     *
     *
     */
    @Override
    public @NonNull Dialog onCreateDialog(Bundle savedInstanceState)
    {
        Log.v(TAG, "show " +   yearNum + " / " + monthNum + " ");

        Activity activity = getActivity();

        // 確認ダイアログの生成
        //final AlertDialog.Builder alertDialog = new AlertDialog.Builder(new ContextThemeWrapper(activity, R.style.wear2_dialog_theme));
        final AlertDialog.Builder alertDialog = new AlertDialog.Builder(activity);

        // Get the layout inflater
        LayoutInflater inflater = activity.getLayoutInflater();
        final View alertView = inflater.inflate(R.layout.target_month_picker, null, false);
        alertDialog.setView(alertView);

        final TextView titleText = alertView.findViewById(R.id.information_picker);
        final NumberPicker month = alertView.findViewById(R.id.number_picker_month);
        final NumberPicker year = alertView.findViewById(R.id.number_picker_year);

        try
        {
            if (title != null)
            {
                titleText.setText(title);
            }
            month.setMinValue(MONTH_LIMIT_MIN);
            month.setMaxValue(MONTH_LIMIT_MAX);
            year.setMinValue(YEAR_LIMIT_MIN);
            year.setMaxValue(YEAR_LIMIT_MAX);

            month.setValue(monthNum);
            year.setValue(yearNum);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }

        alertDialog.setCancelable(true);

        // ボタンを設定する（実行ボタン）
        alertDialog.setPositiveButton(activity.getString(R.string.dialog_positive_execute),
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which)
                    {
                        try
                        {
                            Log.v(TAG, "ENTRY [" + year.getValue() + " / " + month.getValue() + "] ");
                            callback.dataSetYearMonth(year.getValue(), month.getValue());
                        }
                        catch (Exception e)
                        {
                            e.printStackTrace();
                            callback.dataSetCancelled();
                        }
                        dialog.dismiss();
                    }
                });

        // ボタンを設定する (キャンセルボタン）
        alertDialog.setNegativeButton(activity.getString(R.string.dialog_negative_cancel),
                new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int which)
                    {
                        try
                        {
                            callback.dataSetCancelled();
                        }
                        catch (Exception e)
                        {
                            e.printStackTrace();
                        }
                        dialog.cancel();
                    }
                });

        myDialog = alertDialog.create();
        return (myDialog);
    }

    /**
     *
     *
     */
    @Override
    public void onPause()
    {
        super.onPause();
        Log.v(TAG, "AlertDialog::onPause()");
        if (myDialog != null)
        {
            myDialog.cancel();
        }
    }

    /**
     *  コールバックインタフェースの定義
     *
     */
    public interface Callback
    {
        void dataSetYearMonth(int year, int month); // OKを選択したとき
        void dataSetCancelled();  // キャンセルしたとき
    }
}

File: app/src/main/java/net/osdn/gokigen/pkremote/calendar/TargetMonthSetDialog.kt
package net.osdn.gokigen.pkremote.calendar

import android.app.Activity
import android.app.AlertDialog
import android.app.Dialog
import android.content.Context
import android.os.Bundle
import android.util.Log
import android.widget.NumberPicker
import android.widget.TextView
import androidx.fragment.app.DialogFragment
import net.osdn.gokigen.pkremote.R

class TargetMonthSetDialog : DialogFragment()
{
    private var dateCallback: DateCallback? = null

    override fun onAttach(context: Context)
    {
        super.onAttach(context)

        // 呼び出し元の Fragment または Activity から Callback を取得する
        val parentFragment = getParentFragment()
        if (parentFragment is DateCallback)
        {
            dateCallback = parentFragment as DateCallback
        }
        else if (context is DateCallback)
        {
            dateCallback = context as DateCallback
        }
        else
        {
            Log.w(TAG, "Do not implement callback in the caller(Activity/Fragment)")
        }
    }

    override fun onDetach()
    {
        super.onDetach()
        dateCallback = null
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog
    {
        // Bundle から引数を取得（デフォルト値を設定）
        val args = arguments
        val title = if (args != null) {
            args.getString(KEY_TITLE, "")
        } else ""
        val yearNum = args?.getInt(KEY_YEAR, 2026) ?: 2026
        val monthNum = args?.getInt(KEY_MONTH, 1) ?: 1

        Log.v(TAG, "show $yearNum / $monthNum")

        val activity: Activity = requireActivity()
        val alertDialog = AlertDialog.Builder(activity)

        val inflater = activity.layoutInflater
        val alertView = inflater.inflate(R.layout.target_month_picker, null, false)
        alertDialog.setView(alertView)

        val titleText = alertView.findViewById<TextView?>(R.id.information_picker)
        val month = alertView.findViewById<NumberPicker?>(R.id.number_picker_month)
        val year = alertView.findViewById<NumberPicker?>(R.id.number_picker_year)

        if (titleText != null && title != null) {
            titleText.text = title
        }

        if (month != null) {
            month.setMinValue(MONTH_LIMIT_MIN)
            month.setMaxValue(MONTH_LIMIT_MAX)
            month.value = monthNum
        }

        if (year != null) {
            year.setMinValue(YEAR_LIMIT_MIN)
            year.setMaxValue(YEAR_LIMIT_MAX)
            year.value = yearNum
        }

        // 実行ボタンの設定
        alertDialog.setPositiveButton(
            R.string.dialog_positive_execute
        ) { _, _ ->
            if (dateCallback != null && year != null && month != null) {
                Log.v(TAG, "ENTRY [" + year.value + " / " + month.value + "]")
                dateCallback?.dataSetYearMonth(year.value, month.value)
            }
        }

        // キャンセルボタンの設定
        alertDialog.setNegativeButton(R.string.dialog_negative_cancel) { _, _ -> dateCallback?.dataSetCancelled() }

        return alertDialog.create()
    }

    /**
     * コールバックインタフェースの定義
     */
    interface DateCallback
    {
        // OKを選択したとき
        fun dataSetYearMonth(year: Int, month: Int)
        // キャンセルしたとき
        fun dataSetCancelled()
    }

    companion object {
        private val TAG: String = TargetMonthSetDialog::class.java.getSimpleName()

        private const val KEY_TITLE = "key_title"
        private const val KEY_YEAR = "key_year"
        private const val KEY_MONTH = "key_month"

        private const val YEAR_LIMIT_MIN = 2010
        private const val YEAR_LIMIT_MAX = 2050
        private const val MONTH_LIMIT_MIN = 1
        private const val MONTH_LIMIT_MAX = 12

        // インスタンス生成用ファクトリメソッド
        @JvmStatic
        fun newInstance(title: String?, year: Int, month: Int): TargetMonthSetDialog {
            val instance = TargetMonthSetDialog()

            // パラメータはすべて Bundle にまとめて保持させる
            val arguments = Bundle()
            arguments.putString(KEY_TITLE, title)
            arguments.putInt(KEY_YEAR, year)
            arguments.putInt(KEY_MONTH, month)
            instance.setArguments(arguments)

            return instance
        }
    }
}
